scripts/install-multi-user.sh

```diff
@@ -349,7 +349,12 @@ _sudo() {
 }
 
 
-readonly SCRATCH=$(mktemp -d "${TMPDIR:-/tmp/}tmp.XXXXXXXXXX")
+# Ensure that $TMPDIR exists if defined.
+if [[ -n $TMPDIR ]]; then
+    mkdir -m 0700 -p "$TMPDIR"
```

Comment:
I think this needs to either use an intermediate variable or use the same `${TMPDIR:-/tmp}` expansion to avoid this unbound variable error:

https://github.com/abathur/nix/runs/7883911756?check_suite_focus=true#step:4:90

Reply:
I've made a few changes: Use `mktemp -d` without any arguments, and use `[[ -v TMPDIR ]]`.

Reply:
I've re-started CI with these in https://github.com/abathur/nix/actions/runs/2878089386

https://github.com/NixOS/nix/pull/6652 also documents what you need to set up the installer tests on your fork, should you want :)

Reply:
Still failed (but new failure) at https://github.com/abathur/nix/runs/7889093146?check_suite_focus=true#step:4:90

Reply:
Good catch, Bash on macOS doesn't support -v:

```
bash-3.2$ if [[ -v OENUTHOEU ]]; then echo hi ; fi
bash: conditional binary operator expected
bash: syntax error near `OENUTHOEU'

bash-3.2$ if [[ -v $OENUTHOEU ]]; then echo hi ; fi
bash: conditional binary operator expected
bash: syntax error near `$OENUTHOEU'

bash-3.2$ help test
test: test [expr]
    Exits with a status of 0 (true) or 1 (false) depending on
    the evaluation of EXPR.  Expressions may be unary or binary.  Unary
    expressions are often used to examine the status of a file.  There
    are string operators as well, and numeric comparison operators.
    
    File operators:
    
        -a FILE        True if file exists.
        -b FILE        True if file is block special.
        -c FILE        True if file is character special.
        -d FILE        True if file is a directory.
        -e FILE        True if file exists.
        -f FILE        True if file exists and is a regular file.
        -g FILE        True if file is set-group-id.
        -h FILE        True if file is a symbolic link.
        -L FILE        True if file is a symbolic link.
        -k FILE        True if file has its `sticky' bit set.
        -p FILE        True if file is a named pipe.
        -r FILE        True if file is readable by you.
        -s FILE        True if file exists and is not empty.
        -S FILE        True if file is a socket.
        -t FD          True if FD is opened on a terminal.
        -u FILE        True if the file is set-user-id.
        -w FILE        True if the file is writable by you.
        -x FILE        True if the file is executable by you.
        -O FILE        True if the file is effectively owned by you.
        -G FILE        True if the file is effectively owned by your group.
        -N FILE        True if the file has been modified since it was last read.
    
      FILE1 -nt FILE2  True if file1 is newer than file2 (according to
                       modification date).
    
      FILE1 -ot FILE2  True if file1 is older than file2.
    
      FILE1 -ef FILE2  True if file1 is a hard link to file2.
    
    String operators:
    
        -z STRING      True if string is empty.
    
        -n STRING
        STRING         True if string is not empty.
    
        STRING1 = STRING2
                       True if the strings are equal.
        STRING1 != STRING2
                       True if the strings are not equal.
        STRING1 < STRING2
                       True if STRING1 sorts before STRING2 lexicographically.
        STRING1 > STRING2
                       True if STRING1 sorts after STRING2 lexicographically.
    
    Other operators:
    
        -o OPTION      True if the shell option OPTION is enabled.
        ! EXPR         True if expr is false.
        EXPR1 -a EXPR2 True if both expr1 AND expr2 are true.
        EXPR1 -o EXPR2 True if either expr1 OR expr2 is true.
    
        arg1 OP arg2   Arithmetic tests.  OP is one of -eq, -ne,
                       -lt, -le, -gt, or -ge.
    
    Arithmetic binary operators return true if ARG1 is equal, not-equal,
    less-than, less-than-or-equal, greater-than, or greater-than-or-equal
    than ARG2.
```

Reply:
Looks like we probably want -z with bash's default-if-unset syntax:

```
bash-3.2$ set -u
bash-3.2$ if [[ -z "${OENUTHOEU:-}" ]]; then echo hi ; fi
hi
```

Reply:
Started a fresh CI job https://github.com/abathur/nix/actions/runs/2879326487

Reply:
Advanced a line, at least: https://github.com/abathur/nix/runs/7890927130?check_suite_focus=true#step:4:91

I guess the unbound checks are parse-time. Since we're already checking that it isn't empty, I guess something like https://github.com/abathur/nix/commit/9250e20274dcc81055c78c288ecfe14c49728c35 is fine. 

New ci: https://github.com/abathur/nix/actions/runs/2879566885

Reply:
Agh :)

`-v` -> `-z` inverted the condition (variable is set to string is empty). 

Running again with `-n`: https://github.com/abathur/nix/actions/runs/2879795805

Reply:
macOS case still fell over at https://github.com/abathur/nix/runs/7892112794?check_suite_focus=true#step:4:90

It looks like this is a difference in the mkdir implementations, seemingly caused by the `-m` flag:

```console
$ mkdir -m 0700 -p "${TMPDIR:-}"

$ /bin/mkdir -p $TMPDIR

$ /bin/mkdir -m 0700 -p "${TMPDIR:-}"
mkdir: /var/folders/hg/pt5wsk1n54d1prjyg9pff5_r0000gn/T/: Operation not permitted
```

I guess the right next step is to look before we leap. Last swing for the night: https://github.com/abathur/nix/actions/runs/2880189989

Reply:
(this one ran clean)